app/org/sagebionetworks/bridge/upload/StrictValidationHandler.java

```diff
@@ -0,0 +1,233 @@
+        errorList.forEach(context::addMessage);
+
+        // log warning
+        String combinedErrorMessage = "Strict upload validation error in study " + context.getStudy().getIdentifier()
+                + ", upload " + context.getUpload().getUploadId() + ": " + ERROR_MESSAGE_JOINER.join(errorList);
+        logger.warn(combinedErrorMessage);
+
+        // throw error, if configured to do so
+        if (shouldThrow(context.getStudy())) {
+            throw new UploadValidationException(combinedErrorMessage);
+        }
+    }
+
+    /**
+     * <p>
+     * Returns whether strict validation should throw exceptions, based on study configs.
+     * </p>
+     * <p>
+     * This is protected so that unit tests can mock this out. (Until we implement the actual study config lookup.)
+     * </p>
+     *
+     * @param studyIdentifier
+     *         study ID of the current data record we're validating
+     * @return true if we should throw an exception, false otherwise
+     */
+    // TODO: implement this
+    protected boolean shouldThrow(StudyIdentifier studyIdentifier) {
+        return false;
+    }
+
+    /**
+     * Given the schema, the attachments (all we need are names), and the JSON data nodes, we validate the data against
+     * the schema.
+     *
+     * @param fieldDefList
+     *         list of field definitions from the schema
+     * @param attachmentFieldNameSet
+     *         set of attachment field names that we have attachments for
+     * @param recordDataNode
+     *         JSON node of the parsed data to validate
+     * @return list of error messages, empty if there are no errors
+     */
+    private static List<String> validateAllFields(@Nonnull List<UploadFieldDefinition> fieldDefList,
+            @Nonnull Set<String> attachmentFieldNameSet, @Nonnull JsonNode recordDataNode) {
+        // walk the field definitions and validate fields
+        List<String> errorList = new ArrayList<>();
+        for (UploadFieldDefinition oneFieldDef : fieldDefList) {
+            String fieldName = oneFieldDef.getName();
+            UploadFieldType fieldType = oneFieldDef.getType();
+            boolean isRequired = oneFieldDef.isRequired();
+
+            if (UploadFieldType.ATTACHMENT_TYPE_SET.contains(fieldType)) {
+                // For attachment types, since they just get exported as raw files, we only need to check if it's
+                // required and present. Specifically, if it's required and it's not present, then that's an error.
+                if (isRequired && !attachmentFieldNameSet.contains(fieldName)) {
+                    errorList.add("Required attachment field " + fieldName + " missing");
+                }
+            } else {
+                JsonNode fieldValueNode = recordDataNode.get(fieldName);
+
+                if (fieldValueNode != null && !fieldValueNode.isNull()) {
+                    boolean isValid = validateJsonField(fieldType, fieldValueNode);
```

Comment:
Functionally, it's all the same in byte code.

Stylistically, I prefer that if there's a function call that does a lot of processing, it shouldn't be mixed with other things like if statements. I'd prefer to leave this as is, unless you have strong feelings about this.